internal/zinc-persist/src/main/scala/sbt/internal/inc/consistent/ParallelGzipOutputStream.scala

```diff
@@ -0,0 +1,126 @@
+package sbt.internal.inc.consistent
+
+import java.io.{ ByteArrayOutputStream, FilterOutputStream, OutputStream }
+import java.util.zip.{ CRC32, Deflater, DeflaterOutputStream }
+
+import scala.annotation.tailrec
+import scala.concurrent.duration.Duration
+import scala.concurrent.{ Await, ExecutionContext, Future }
+import scala.collection.mutable
+
+/**
+ * Parallel gzip compression. Algorithm based on https://github.com/shevek/parallelgzip
+ * with additional optimization and simplification. This is essentially a block-buffered
+ * stream but instead of writing a full block to the underlying output, it is passed to a
+ * thread pool for compression and the Futures of compressed blocks are collected when
+ * flushing.
+ */
+object ParallelGzipOutputStream {
+  private val blockSize = 64 * 1024
+  private val compression = Deflater.DEFAULT_COMPRESSION
+
+  private class BufOut(size: Int) extends ByteArrayOutputStream(size) {
+    def writeTo(buf: Array[Byte]): Unit = System.arraycopy(this.buf, 0, buf, 0, count)
+  }
+
+  private class Worker {
+    private[this] val defl = new Deflater(compression, true)
+    private[this] val buf = new BufOut(blockSize + (blockSize >> 3))
+    private[this] val out = new DeflaterOutputStream(buf, defl, true)
+    def compress(b: Block): Unit = {
+      defl.reset()
+      buf.reset()
+      out.write(b.data, 0, b.length)
+      out.flush()
+      b.length = buf.size
+      if (b.length > b.data.length) b.data = new Array[Byte](b.length)
+      buf.writeTo(b.data)
+    }
+  }
+
+  private val localWorker = new ThreadLocal[Worker] {
+    override def initialValue = new Worker
+  }
+
+  private class Block {
+    var data = new Array[Byte](blockSize + (blockSize >> 3))
+    var length = 0
+  }
+
+  private val header = Array[Byte](0x1f.toByte, 0x8b.toByte, Deflater.DEFLATED, 0, 0, 0, 0, 0, 0, 0)
+}
+
+final class ParallelGzipOutputStream(
+    out: OutputStream,
+    ec: ExecutionContext = ExecutionContext.global, // NOLINT
+    parallelism: Int = Runtime.getRuntime.availableProcessors()
```

Comment:
Could we remove the default arguments here to avoid inadvertently use `global`?

Reply:
We *are* using `global. Is there a better EC that is accessible when you construct an AnalysisStore? Should this be a parameter of the AnalysisStore.text/binary methods?

Reply:
I'm moving the implicits up into the ConsistentAnalysisStore factory methods and making them explicit further down.